### PR TITLE
don't re-initialize RTC if it survived a reboot (partial)

### LINKED ISSUE #160: RTC should optionally reset the backup domain
I've got an F0 board with a battery backup on the RTC, so it would be convenient to have an RTC class that only resets the backup domain if it has not already been configured. In the STM32 example this is detected by reading a magic value from a backup register `RTC_ReadBackupRegister(RTC_BKP_DR0) != BKP_VALUE` (rather than a dedicated bit for this function that just detects if the calendar year is not 0...)

One strategy that I will try in my fork might be to have RtcBase check the backup register, add a flag on RtcBase, feature class resets or not based on flag. However as it is currently structured the magic number could not be a template argument? What if you want to reset the RTC again when the user sets the time just in case?

### PR DESCRIPTION
**Pull request by dholth**
Here's a concept implementation of the feature. If the magic value exists in the backup register, then we assume the RTC is still correctly configured. Based on the stdperiph example. Fixes #160 

Not tested yet and not completely elegant. I would like to have a way to change the BKP_VALUE to change so new firmware has a chance to reset the clock and I'm not entirely comfortable storing extra data in _hourFormat.

### CODE REVIEW

File: lib/include/rtc/f0/RtcBase.h [hunk 1]
```diff
@@ -48,10 +54,14 @@ namespace stm32plus {
...
-    // reset the backup domain
+    if (RTC_ReadBackupRegister(RTC_BKP_DR0) != BKP_VALUE) {
+      // reset the backup domain
+      RCC_BackupResetCmd(ENABLE);
+      RCC_BackupResetCmd(DISABLE);
 
-    RCC_BackupResetCmd(ENABLE);
-    RCC_BackupResetCmd(DISABLE);
+      RTC_WriteBackupRegister(RTC_BKP_DR0, BKP_VALUE);
+      _hourFormat |= SURVIVED_FLAG;
```
Comment (dholth) [inline lib/include/rtc/f0/RtcBase.h]: Oops, this is 'did not survive'

**Comment by dholth** [comment]
I tried the second example of my board and the CR1225 battery-backed LSE works (with the 10-second blink example from stm32plus). It does work after reboot without being re-initialized. I thought one way to solve the "do I want unconditional reset?" problem would be to pass the magic backup value to RtcBase but this is unresolved.

**Comment by dholth** [comment]
It works. If I want to persist the RTC, I put a battery in my board and I instantiate the RTC with a magic number. Rtc<RtcLseClockFeature> rtc(RTC_HourFormat_24, 0x12345678);

Unrelated bug, if the LSE is broken, while(RCC_GetFlagStatus(RCC_FLAG_LSERDY)==RESET); in RtcLseClockFeature doesn't time out and the program hangs.

**Comment by dholth** [comment]
If you are happy with this approach I'll edit the other chips to work the same way, and will also move the writing of the backup register to the Rtc() constructor after all of the feature classes have initialized.

**Comment by dholth** [comment]
F0 reference manual section 7.2.9:

"The LSE clock is in the Backup domain, whereas the HSE and LSI clocks are not. Consequently:
If LSE is selected as RTC clock:
The RTC continues to work even if the VDD supply is switched off, provided the VBAT supply is maintained.
LSI is selected as the RTC clock: The RTC state is not guaranteed if the VDD supply is powered off.
The HSE clock divided by 32 is used as the RTC clock: The RTC state is not guaranteed if the VDD supply is powered off or if the internal voltage regulator is powered off (removing power from the 1.8 V domain)."

**Comment by andysworkshop** [comment]
That explains why my test with the LSI didn't work.

**Pull request closed without merging**